### PR TITLE
SAM-2479; map /direct/sam_pub when copying sites

### PR DESCRIPTION
**Pull request by clhedrick**

### CODE REVIEW

**Comment by clhedrick** [comment]
References to sam_pub/NNN get mapped to sam_core/NNN in the new site. sam_core refers to the most recent assessment published from that core assessment. In case someone copies a site that has been copied and so has /direct/sam_core, I also map sam_core references to the new ones.

The other possible implementation would be to add a small table (or a field to the base assessments table) where we could record the published assessment ID the first time one is found. That way there's no danger of having the reference change if a new assessment is published from the same core assessment. It's unclear which behavior instructors would expect. I sort of think using the most recent may actually be right.

Note that the published assessments table is already indexed by the base assessment ID, so there's very little overhead to this code. One additional indexed looking when evaluating the link.

**Comment by clhedrick** [comment]
Note that the code in the migration helper has two sakai.properties values, one controlling which items have [] around them and the other which URLs are turned into text. If a site for some reason doesn't like this mapping, they can maintain the current behavior by configuration.

File: samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/entity/api/CoreAssessmentEntityProvider.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.sakaiproject.tool.assessment.entity.api;
+
+import org.sakaiproject.entitybroker.entityprovider.EntityProvider;
+
+/**
+ * Entity Provider for samigo PublishedAssessments
+ * 
+ * @author Joshua Ryan  josh@asu.edu  alt^I
```
Comment (ottenhoff) [inline samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/entity/api/CoreAssessmentEntityProvider.java]: This author line seems copied and pasted.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/rsf/bean/PublishedAssessmentVPInferrer.java [hunk 2]
```diff
@@ -16,14 +17,37 @@
...
 
+  private PublishedAssessmentFacadeQueriesAPI publishedAssessmentFacadeQueries;
+  public void setPublishedAssessmentFacadeQueries(
+      PublishedAssessmentFacadeQueriesAPI publishedAssessmentFacadeQueries) {
+    this.publishedAssessmentFacadeQueries = publishedAssessmentFacadeQueries;
+  }
+
   public String[] getHandledPrefixes() {
-    return new String[] {PublishedAssessmentEntityProvider.ENTITY_PREFIX};
+      return new String[] {PublishedAssessmentEntityProvider.ENTITY_PREFIX, "sam_core"};
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/rsf/bean/PublishedAssessmentVPInferrer.java]: Can you use the ENTITY_PREFIX  constant instead?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/rsf/bean/PublishedAssessmentVPInferrer.java [hunk 3]
```diff
@@ -16,14 +17,37 @@
...
   }
 
   public ViewParameters inferDefaultViewParameters(String reference) {
     BeginAssessmentViewParameters params = new BeginAssessmentViewParameters();
     IdEntityReference ep = new IdEntityReference(reference);
-    params.pubReference = ep.id;
+    if (reference.startsWith("/sam_core/")) {
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/rsf/bean/PublishedAssessmentVPInferrer.java]: Can the ENTITY_PREFIX  constant be used instead?

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/entity/impl/CoreAssessmentEntityProviderImpl.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import java.lang.IllegalStateException;
+
+
+
+/**
+ * Entity Provider impl for samigo PublishedAssessments
+ * 
+ * Implements PropertyProvideable to allow things to browse the entities provided by site
+ * 
+ * @author Joshua Ryan  josh@asu.edu  alt^I
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/entity/impl/CoreAssessmentEntityProviderImpl.java]: Copy/paste

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java [hunk 5]
```diff
@@ -1840,6 +1843,7 @@ public void copyAllAssessments(String fromContext, String toContext) {
 			AssessmentData new_a = prepareAssessment(a,
 					ServerConfigurationService.getServerUrl(), toContext);
 			newList.add(new_a);
+			assessmentMap.put(new_a, "sam_core/" + a.getAssessmentBaseId());
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java]: Can the constant be used here?

**Comment by ern** [comment]
FYI I noticed in this PR there is a commit from master this needs to be removed...

It looks like you merged from master when you should have rebased from master?

**Comment by clhedrick** [comment]
No. I did a commit to that branch by mistake when I meant to do it to my own project. When I deal with the other suggestions I'll fix that, one way or another.

**Pull request closed without merging**